src/core/server/workspaces/saved_objects/workspace_saved_objects_client_wrapper.ts

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import {
+  OpenSearchDashboardsRequest,
+  SavedObjectsClientWrapperFactory,
+  SavedObjectsCreateOptions,
+  SavedObjectsDeleteOptions,
+} from 'opensearch-dashboards/server';
+import {
+  WorkspacePermissionControl,
+  WorkspacePermissionMode,
+} from '../workspace_permission_control';
+
+const getWorkspaceIdFromRequest = (request: OpenSearchDashboardsRequest) => {
+  if (typeof request.headers.referer !== 'string') {
+    return null;
+  }
+  const refererURL = new URL(request.headers.referer);
+  return /^\/w\/([^\/]+)/.exec(refererURL.pathname)?.[1];
+};
+
+export class WorkspaceSavedObjectsClientWrapper {
+  private async validatePermissionByRequest(
+    request: OpenSearchDashboardsRequest,
+    permissionMode: WorkspacePermissionMode | WorkspacePermissionMode[]
+  ) {
+    const requestWorkspaceId = getWorkspaceIdFromRequest(request);
+    if (
+      requestWorkspaceId &&
+      !(await this.permissionControl.validate(requestWorkspaceId, permissionMode, request))
+    ) {
+      throw new Error('Invalidate workspace permission');
+    }
+  }
+  public wrapperFactory: SavedObjectsClientWrapperFactory = (wrapperOptions) => {
+    const deleteWithWorkspacePermissionControl = async (
+      type: string,
+      id: string,
+      options: SavedObjectsDeleteOptions = {}
+    ) => {
+      await this.validatePermissionByRequest(wrapperOptions.request, WorkspacePermissionMode.Admin);
+      return await wrapperOptions.client.delete(type, id, options);
+    };
+
+    const createWithWorkspacePermissionControl = async <T = unknown>(
+      type: string,
+      attributes: T,
+      options?: SavedObjectsCreateOptions
+    ) => {
+      await this.validatePermissionByRequest(wrapperOptions.request, WorkspacePermissionMode.Admin);
```

Comment:
We seem need to interrupt or modify the response body if verification result is not successful. But as i see the only example from `src/plugins/data_source/server/saved_objects/data_source_saved_objects_client_wrapper.ts`, they seem use `wrapperFactory` to distinguish request params. What's your opinion?